tempesta_fw/sock_clnt.c

```diff
@@ -123,15 +121,21 @@ tfw_sock_clnt_new(struct sock *sk)
 		goto err_conn_init;
 	}
 
+	spin_lock(&conn->splock);
+	tfw_connection_link_sk(conn, sk);
+	tfw_connection_link_peer(conn, (TfwPeer *)cli);
+	spin_unlock(&conn->splock);
+
+	ss_set_callbacks(sk);
+
 	TFW_DBG("new client socket is accepted: sk=%p, conn=%p, cli=%p\n",
 		sk, conn, cli);
 	return 0;
 
 err_conn_init:
-	tfw_connection_unlink_peer(conn);
-	tfw_connection_unlink_sk(conn);
 	tfw_connection_destruct(conn);
-	tfw_cli_conn_free(conn);
+	if (tfw_connection_put(conn))
+		tfw_cli_conn_free(conn);
```

Comment:
Double connection freeing. The second freeing is at tfw_connection_destruct() -> conn_destruct hook  which is tfw_http_conn_destruct() -> tfw_http_conn_msg_free() -> tfw_connection_put() & tfw_cli_conn_free()